lib/verk/queue_stats_counters.ex

```diff
@@ -0,0 +1,70 @@
+defmodule Verk.QueueStatsCounters do
+  @moduledoc """
+  This module is responsible for abstracting the logic of keeping counters for
+  each queue.
+  """
+
+  @counters :queue_stats
+  @ets_options [:ordered_set, :named_table, read_concurrency: true, keypos: 1]
+
+  @doc """
+  Initializes the ets tables for the queue stats.
+  """
+  @spec init :: :ok
+  def init do
+    :ets.new(@counters, @ets_options)
+  end
+
+  @doc """
+  Updates the counters according to the event that happened.
+  """
+  @spec register(:started | :finished | :failed, binary) :: integer
+  def register(:started, queue), do: update_counters(queue, { 2, 1 })
+  def register(:finished, queue), do: update_counters(queue, [{ 3, 1 }, { 2, -1 }])
+  def register(:failed, queue), do: update_counters(queue, [{ 4, 1 }, { 2, -1 }])
+
+  @doc """
+  Saves processed and failed total counts to Redis.
+  """
+  @spec persist :: {:ok, [Redix.Protocol.redis_value]} | {:error, atom}
+  def persist do
+    {total_processed, total_failed} = Enum.reduce stats, {0, 0}, fn {queue, processed, failed}, {accum_processed, accum_failed} ->
+      redis_commands = [
+        incrby(queue, :processed, processed),
+        incrby(queue, :failed, failed)
+      ]
+      case Redix.pipeline(Verk.Redis, redis_commands) do
+        {:ok, [%Redix.Error{}, %Redix.Error{}]} ->
+          {accum_processed, accum_failed}
+        {:ok, [_, %Redix.Error{}]} ->
+          update_counters(queue, { 3, -processed })
+          {accum_processed + processed, accum_failed}
+        {:ok, [%Redix.Error{}, _]} ->
+          update_counters(queue, { 4, -failed })
+          {accum_processed, accum_failed + failed}
+        {:ok, [_, _]} ->
+          update_counters(queue, [{ 3, -processed }, { 4, -failed }])
+          {accum_processed + processed, accum_failed + failed}
+        _error ->
+          {accum_processed, accum_failed}
+      end
+    end
+
+    redis_commands = [
+      incrby(:total, :processed, total_processed),
+      incrby(:total, :failed, total_failed)
+    ]
+    Redix.pipeline(Verk.Redis, redis_commands)
```

Comment:
I am doing that but in a different way. I think there are two problems with totally cleaning up the table. The first is that we don't get a success or error for every redis command in the pipeline but to each one of them. The other problem is that due concurrency, there can be event notifications while the `persist` function is being run. If we clean up the table we might lose these jobs. That's why I am decrementing each of the counters by the value I successfully placed in Redis.

```elixir
  update_counters(queue, [{ 3, -processed }, { 4, -failed }])
  {accum_processed + processed, accum_failed + failed}
```

This case is the one, both of the commands are successful, but you may see the ones above for error cases. After the counters for each queue is decremented and accumulated in the total, there is no more need to clean up anything. This way we are safe for the two problems I mentioned before. Do you think there's a better way to do it?